### PR TITLE
Add checks to allow logging only specific datatypes

### PR DESCRIPTION
**Pull request by HallikiM**
Also adding tests

---
Currently, it allows logging of objects and the whitelisting does not apply to the fields of that object.

This change adds data type checks and allows logging of only specific types so that data that is not whitelisted or that is blacklisted does not go through. This means it will be either masked or pruned.

MSG-709

### CODE REVIEW

**Comment by kmteras** [comment]
You linked your team's caretaker journal. Could you also link the incident ticket (IN-299) or create an internal ticket for this problem for tracking fixes as well?

**Comment by indrekj** [comment]
Tests please as well

**Comment by HallikiM** [comment]
> You linked your team's caretaker journal. Could you also link the incident ticket (IN-299) or create an internal ticket for this problem for tracking fixes as well?

linked the incident ticket

**Review by kmteras — CHANGES_REQUESTED** [review]
Changes look fine but please add the tests

File: spec/lenjador/preprocessors/whitelist_spec.rb [hunk 1]
```diff
@@ -312,8 +387,92 @@ def process(pointers, data)
...
     end
   end
 end
+
+class Test
+  attr_accessor :asd
+
+  def initialize
+    @asd = 'tdlgdfaha'
+  end
```
Comment (kmteras) [inline spec/lenjador/preprocessors/whitelist_spec.rb:477]: Random fields, I know, but could name them something nicer 😄

**Review by kmteras — APPROVED** [review]

File: spec/lenjador/preprocessors/whitelist_spec.rb [hunk 2]
```diff
@@ -312,8 +387,92 @@ def process(pointers, data)
...
+      end
+    end
+
     def process(pointers, data)
       described_class.new(pointers: pointers, action: :prune).process(data)
     end
   end
 end
+
+class Test
```
Comment (indrekj) [inline spec/lenjador/preprocessors/whitelist_spec.rb:472]: This is a bad name to have globally defined. You don't need a global class, you can do:
```
let(:data) do
  Class.new do
    attr_accessor :some_attribute
    #...
  end.new
end
```
I personally would use an open struct instead:
```
let(:data) { OpenStruct.new(:name => "Rowdy", :owner => "John Smith") }
```

File: lib/lenjador/preprocessors/strategies/prune.rb [hunk 3]
```diff
@@ -18,8 +18,11 @@ def process(data, pointer = '')
           when Array
             process_array(data, pointer)
 
-          else
+          when String, Numeric, Symbol, Date, Time, TrueClass, FalseClass, NilClass
```
Comment (indrekj) [inline lib/lenjador/preprocessors/strategies/prune.rb:21]: `Date` is not loaded by default.

```
 ~ irb
irb(main):001:0> Date
Traceback (most recent call last):
        4: from /home/indrek/.asdf/installs/ruby/2.7.8/bin/irb:23:in `<main>'
        3: from /home/indrek/.asdf/installs/ruby/2.7.8/bin/irb:23:in `load'
        2: from /home/indrek/.asdf/installs/ruby/2.7.8/lib/ruby/gems/2.7.0/gems/irb-1.2.6/exe/irb:11:in `<top (required)>'
        1: from (irb):1
NameError (uninitialized constant Date)
```
And if it's loaded, then there's also `DateTime`.

Comment (HallikiM) [reply lib/lenjador/preprocessors/strategies/prune.rb:21]: Added tests for this as well. DateTime is subclass of Date.

**Comment by indrekj** [comment]
@HallikiM https://github.com/salemove/lenjador/pull/46/files#r1425454770

**Review by indrekj — APPROVED** [review]

**Pull request merged**